server/src/main/java/org/elasticsearch/action/admin/indices/settings/get/GetSettingsResponse.java

```diff
@@ -71,5 +118,121 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeString(cursor.key);
             Settings.writeSettingsToStream(cursor.value, out);
         }
+        if (out.getVersion().onOrAfter(org.elasticsearch.Version.V_7_0_0_alpha1)) {
+            out.writeVInt(indexToDefaultSettings.size());
+            for (ObjectObjectCursor<String, Settings> cursor : indexToDefaultSettings) {
+                out.writeString(cursor.key);
+                Settings.writeSettingsToStream(cursor.value, out);
+            }
+        }
+    }
+
+    private static void parseSettingsField(XContentParser parser, String currentIndexName, Map<String, Settings> indexToSettings,
+                                           Map<String, Settings> indexToDefaultSettings) throws IOException {
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser::getTokenLocation);
+        String settingsFieldName = parser.currentName(); //this should be either "settings" or "defaults"
+        parser.nextToken(); //go to settings object itself
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+
+        switch (settingsFieldName) {
+            case "settings":
+                indexToSettings.put(currentIndexName, Settings.fromXContent(parser));
+                break;
+            case "defaults":
+                indexToDefaultSettings.put(currentIndexName, Settings.fromXContent(parser));
+                break;
+            default:
+                //We don't know this field, ignore it
+        }
+    }
+
+    public static GetSettingsResponse fromXContent(XContentParser parser) throws IOException {
+        HashMap<String, Settings> indexToSettings = new HashMap<>();
+        HashMap<String, Settings> indexToDefaultSettings = new HashMap<>();
+
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+        parser.nextToken();
+
+        while (!parser.isClosed() && parser.currentToken() == XContentParser.Token.FIELD_NAME) {
+            String indexName = parser.currentName();
+            parser.nextToken(); //go to per-index settings object
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
```

Comment:
instead of making sure that we find is an object, shall we rather ignore anything that's not an object at this level?